### PR TITLE
C#: Add support for the tool status page

### PR DESCRIPTION
**Pull request by mbg**
This is the same as #12217 which was reverted by #12379, with the addition of 9dc9925f59ea93da74dfbfc570bedd70d35dddcd which avoids creating files for the diagnostics in the unit tests altogether, therefore avoiding potential race conditions in the tests.

### CODE REVIEW

**Review by michaelnebel — APPROVED** [review]
Looks plausible to me!

File: csharp/extractor/Semmle.Util/ToolStatusPage.cs [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+using Newtonsoft.Json;
+using Newtonsoft.Json.Converters;
+using Newtonsoft.Json.Serialization;
+
+namespace Semmle.Util
+{
+    /// <summary>
+    /// Represents diagnostic messages for the tool status page.
+    /// </summary>
+    public class DiagnosticMessage
```
Comment (tamasvajk) [inline csharp/extractor/Semmle.Util/ToolStatusPage.cs:15]: Why are these classes placed in the same file? (I was surprised to find `DiagnosticMessage` in `ToolStatusPage.cs`.) Should `ToolStatusPage.cs` become a folder instead?

Comment (tamasvajk) [inline csharp/extractor/Semmle.Util/ToolStatusPage.cs:15]: Why are these classes placed in the same file? (I was surprised to find `DiagnosticMessage` in `ToolStatusPage.cs`.) Should `ToolStatusPage.cs` become a folder instead?

File: csharp/autobuilder/Semmle.Autobuild.CSharp/AutoBuildRule.cs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+            var tryCleanExtractorArgsLogs =
+                BuildScript.Create(actions =>
+                {
+                    foreach (var file in Extractor.GetCSharpArgsLogs())
+                    {
+                        try
+                        {
+                            actions.FileDelete(file);
+                        }
+                        catch // lgtm[cs/catch-of-all-exceptions] lgtm[cs/empty-catch-block]
```
Comment (tamasvajk) [inline csharp/autobuilder/Semmle.Autobuild.CSharp/AutoBuildRule.cs:39]: This code was probably only moved. Is there any value in keeping `lgtm[cs/catch-of-all-exceptions] lgtm[cs/empty-catch-block]` around?

File: csharp/autobuilder/Semmle.Autobuild.CSharp/CSharpAutobuilder.cs [hunk 3]
```diff
@@ -29,25 +31,16 @@ public CSharpAutobuildOptions(IBuildActions actions) : base(actions)
 
     public class CSharpAutobuilder : Autobuilder<CSharpAutobuildOptions>
     {
-        public CSharpAutobuilder(IBuildActions actions, CSharpAutobuildOptions options) : base(actions, options) { }
+        private const string buildCommandDocsUrl =
+            "https://docs.github.com/en/code-security/code-scanning/automatically-scanning-your-code-for-vulnerabilities-and-errors/configuring-the-codeql-workflow-for-compiled-languages";
```
Comment (tamasvajk) [inline csharp/autobuilder/Semmle.Autobuild.CSharp/CSharpAutobuilder.cs:35]: Embedding this URL in the code feels rather odd. Let's hope this is a somewhat persistent URL.

File: csharp/autobuilder/Semmle.Autobuild.CSharp/CSharpDiagnosticClassifier.cs [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+            public HashSet<string> MissingProjectFiles { get; }
+
+            public Result()
+            {
+                this.MissingProjectFiles = new HashSet<string>();
+            }
+
+            public DiagnosticMessage ToDiagnosticMessage<T>(Autobuilder<T> builder, DiagnosticMessage.TspSeverity? severity = null) where T : AutobuildOptionsShared => new(
+                builder.Options.Language,
+                "missing-project-files",
```
Comment (tamasvajk) [inline csharp/autobuilder/Semmle.Autobuild.CSharp/CSharpDiagnosticClassifier.cs:76]: How is `no-projects-or-solutions` different from `missing-project-files`?

File: csharp/autobuilder/Semmle.Autobuild.Shared/DiagnosticClassifier.cs [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+using System.Text.RegularExpressions;
+using Semmle.Util;
+
+namespace Semmle.Autobuild.Shared
+{
+    /// <summary>
+    /// Direct results result from the successful application of a <see cref="DiagnosticRule" />,
+    /// which can later be converted to a corresponding <see cref="DiagnosticMessage" />.
+    /// </summary>
+    public interface IDiagnosticsResult
```
Comment (tamasvajk) [inline csharp/autobuilder/Semmle.Autobuild.Shared/DiagnosticClassifier.cs:11]: If there's no strong reason to keep these classes and interfaces in this file, I'd go with separate files for the declarations.

File: csharp/extractor/Semmle.Util/ToolStatusPage.cs [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+
+        /// <summary>
+        /// Stores flags indicating where the diagnostic should be displayed.
+        /// </summary>
+        public class TspVisibility
+        {
+            /// <summary>
+            /// True if the message should be displayed on the status page (defaults to false).
+            /// </summary>
+            public bool? StatusPage { get; }
```
Comment (tamasvajk) [inline csharp/extractor/Semmle.Util/ToolStatusPage.cs:63]: Do these booleans need to be nullable? Their comment say they anyways default to false.

**Review by tamasvajk — COMMENTED** [review]
Added some minor comments and questions.

File: csharp/autobuilder/Semmle.Autobuild.CSharp/CSharpDiagnosticClassifier.cs (hunk 4 above)
Comment (mbg) [reply csharp/autobuilder/Semmle.Autobuild.CSharp/CSharpDiagnosticClassifier.cs:76]: The former means that no project or solution files were found at all. The latter means some were found, but are importing some other project files that do not exist.

File: csharp/extractor/Semmle.Util/ToolStatusPage.cs (hunk 6 above)
Comment (mbg) [reply csharp/extractor/Semmle.Util/ToolStatusPage.cs:63]: There is a difference here. If they are `null`, then they won't get serialised and don't appear in the resulting JSON. The description is taken from the ADR, so it actually defaults to `false` in the CLI (or somewhere else further down the pipeline).

**Pull request merged**